libenv/sysinfo.c

```diff
@@ -2537,24 +2535,42 @@ static void Xen_Cpuid(uint32_t idx, uint32_t *eax, uint32_t *ebx, uint32_t *ecx,
 
 /******************************************************************/
 
-static int Xen_Hv_Check(void)
+static bool Xen_Hv_Check(void)
 {
-    uint32_t eax;
+    /* CPUID interface to Xen from arch-x86/cpuid.h:
+     * Leaf 1 (0x40000000)
+     * EAX: Largest Xen-information leaf. All leaves up to an including @EAX
+     *   are supported by the Xen host.
+     * EBX-EDX: "XenVMMXenVMM" signature, allowing positive identification
+     *   of a Xen host.
+     *
+     * Additional information can be found in the Hypervisor CPUID
+     * Interface Proposal (https://lkml.org/lkml/2008/10/1/246)
+     */
+
+    uint32_t eax, base;
     union
     {
         uint32_t u[3];
         char s[13];
     } sig = {{0}};
 
-    Xen_Cpuid(0x40000000, &eax, &sig.u[0], &sig.u[1], &sig.u[2]);
-
-    if (strcmp("XenVMMXenVMM", sig.s) || (eax < 0x40000002))
+    for (base = 0x40000000; base < 0x40010000; base += 0x100)
```

Comment:
This unfortunately is causing segfaults on our older platforms (RHEL4 in particular). I still don't fully understand the choice of this value, since I can't see it mentioned anywhere, but admittedly I don't know much about Xen detection.

Did you just choose a value that you thought was "high enough"? Do you think it might fix the segfaults if we reduced this value?